### PR TITLE
fill out pointers

### PR DESCRIPTION
**Pull request by dneto0**
Fixes: #16

### CODE REVIEW

File: content/types/pointers/specifying.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+A pointer type is written as
+{{< ptr "ptr<(AS),(T),(AM)>" >}} or {{< ptr "ptr<(AS),(T)>" >}}, where
+* {{< ptr "(AS)" >}} is an address space,
+* {{< ptr "(T)" >}} is a type, known as the store type, and
+* {{< ptr "(AM)" >}} is an access mode.
+
+The access mode can only be written in WGSL when the address space is `storage`,
+and defaults in this case to `read`.
+Otherwise the acces mode takes a [default based on address space](https://w3.org/TR/WGSL#address-space).
```
Comment (dj2) [inline content/types/pointers/specifying.md:24]: I find these two sentences hard to parse.

```
Only the `storage` address space has an access mode. (With a default of `read` if not provided).

Outside of storage there is always an [implicit access mode](...) based on the addresss space.
```

?

File: content/types/pointers/using.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+---
+title: "Using pointers"
+shader: ./using.wgsl
+---
+
+Get a pointer from a variable with the `&` unary operator:
```
Comment (dj2) [inline content/types/pointers/using.md:6]: Drop the `unary` ? Doesn't seem relevant.

File: content/types/pointers/using.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+* If `v` is a variable, then `&v` is a pointer referring to all the memory for `v`.
```
Comment (dj2) [inline content/types/pointers/using.md:7]: `to all the memory` seems weird, techincally it's a pointer to the _start_ of the memory for `v`, no?

Maybe just 'referring to the memory for `v`'?

File: content/types/pointers/using.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+Get a pointer to part of a composite variable by writing an expression to access that part,
+then use the `&` unary operator.
```
Comment (dj2) [inline content/types/pointers/using.md:10]: ditto unary

File: content/types/pointers/using.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+* If `chair` is a structure with member `legs` that is an array of 4 values, then `&chair.legs[3]`
+   is a pointer to the last leg in the chair.
+
+To access the memory a pointer refers to, use the `*` unary operator to turn the pointer into a reference.
```
Comment (dj2) [inline content/types/pointers/using.md:14]: ditto unary

File: content/types/pointers/using.wgsl [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  let particles_ptr = &particles;  // Here is a pointer
+  for (var i: u32; i < arrayLength(particles_ptr); i++) {
+    // Use & to get a pointer as a short name for something inside a big variable.
+    let a_particle = &particles[i];
+
+    // We could have spelled out the pointer type in full.
+    // let a_particle: ptr<storage,Particle,read_write> = &particles[i];
+
+    // Now use 'a_particle' as a ahort name.
+    let its_pos = (*a_particle).pos;
```
Comment (dj2) [inline content/types/pointers/using.wgsl:19]: Are the `()`s required, when as they aren't used in other examples. Maybe a comment?

File: content/types/pointers/using.md (hunk 2 above)
Comment (dneto0) [reply content/types/pointers/using.md:6]: & is also a binary bitwise operator on integers (and bools?)

Comment (dj2) [reply content/types/pointers/using.md:6]: True, but I don't really care. I'm using the `&` operator. It's giving me a pointer when I write it like this. I can use it in other ways and do other things, but the fact that this is `unary` doesn't mean much.

File: content/types/pointers/no_aliasing.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
+---
+title: "No Aliasing Allowed With Pointer Paramters"
+shader: ./no_aliasing.wgsl
+---
+
+A pointer parameter to a user-declared function is not allowed to alias, when
```
[the remark this answers is not in the document]
Comment (dneto0) [reply content/types/pointers/no_aliasing.md:6]: Going with:

> Pointer parameters to user-declared functions are not allowed to alias,
when there is a potential write via the pointer or its alias.

File: content/types/pointers/no_aliasing.wgsl [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+var<private> x : i32 = 0;
+fn set_from_x(p : ptr<private, i32>) { *p = x; }
+fn dont_set_from_x(p : ptr<private, i32>) {
+  if (false) {
+    *p = x; // Still counts as a potential write.
+  }
+}
+
+fn main() {
+  // set_from_x(&x);      // Error: set_from_x would have two views of x.
```
[the remark this answers is not in the document]
Comment (dneto0) [reply content/types/pointers/no_aliasing.wgsl:20]: Rewrote with more words.

File: content/types/pointers/passing_pointers.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
+---
+title: "Pointers as function parameters"
+shader: ./passing_pointers.wgsl
+---
+
+Pointers can be passed into **user-declared** functions, but with restrictions:
```
[the remark this answers is not in the document]
Comment (dneto0) [reply content/types/pointers/passing_pointers.md:6]: arrayLength takes a pointer paramter, and so do a few others (e.g. atomics). 
I'll rewrite.

[the remark this answers is not in the document]
Comment (dneto0) [reply content/types/pointers/passing_pointers.md:6]: Will rewrite.

File: content/types/pointers/passing_pointers.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+1. Only pointers into `function` or `private` address spaces are allowed.
+2. Only pointers to whole variables can be used.
+    * You can't pass the address of only part of a composite variable.
+3. Aliases are not allowed when there is a potential write through one of them.  
+    We'll explain more on the [next page]({{< ref "types/pointers/no_aliasing" >}}).
+
+> These constraints make it safe to translate WGSL pointer parameters as `inout` parameters 
+> in HLSL and GLSL.
+
+Passing pointers into builtin functions has fewer restrictions.
```
[the remark this answers is not in the document]
Comment (dneto0) [reply content/types/pointers/passing_pointers.md:17]: Yeah. Rewriting with more text.

File: content/types/pointers/specifying.md (hunk 1 above)
Comment (dneto0) [reply content/types/pointers/specifying.md:24]: Rewritten.

File: content/types/pointers/using.md (hunk 2 above)
Comment (dneto0) [reply content/types/pointers/using.md:6]: Dropped "unary" in three places.

File: content/types/pointers/using.md (hunk 3 above)
Comment (dneto0) [reply content/types/pointers/using.md:7]: No. In WGSL technically it's a memory view for *all* the memory locations.

In C it's functionally equivalent to the byte address of the first location.

The WGSL definition works with how we make references and pointers duals of each other, and how we specify what memory locations are accessed via reads and writes.

File: content/types/pointers/using.wgsl (hunk 6 above)
Comment (dneto0) [reply content/types/pointers/using.wgsl:19]: Updated with text in the main body. Split this into a new page.

**Review by dj2 — APPROVED** [review]
